Yatima/Typechecker/Infer.lean

```diff
@@ -21,136 +21,125 @@ namespace Yatima.Typechecker
 
 /-- Reduces the application of a `pi` type to its arguments -/
 def applyType : Value → List (Thunk Value) → TypecheckM Value
-  | .pi _ _ _ img imgCtx, arg :: args => do
+  | .pi _ _ _ _ img imgCtx, arg :: args => do
     let res ← withEnv (imgCtx.extendWith arg) (eval img)
     applyType res args
   | type, [] => pure type
   | _, _ => throw .cannotApply
 
--- /-- Checks if a type is an unit inductive -/
--- def isUnit : Value → TypecheckM Bool
---   | .app (.const name i _) _ => do
---     match ← derefConst name i with
---     | .inductive induct => pure induct.unit
---     | _ => pure false
---   | _ => pure false
-
-/-- Determines if a value at level `lvl` is a `Prop` ( ↔ `Sort 0`) -/
-partial def isProp (lvl : Nat) : Value → TypecheckM Bool
-  | .pi name _ dom img ctx => do
-    let res ← withNewExtendedEnvByVar ctx name lvl $ eval img
-    -- A pi type is a proposition if and only if its image is a proposition
-    isProp (lvl + 1) res
-  | .app neu args => do
-    let type ← match neu with
-      | .const constName k us => do
-        let const ← derefConst constName k
-        let ctx := { (← read) with env := ⟨ [], us ⟩ }
-        pure $ suspend const.type ctx
-      | .fvar _ _ => sorry
-      | .proj _ _ => sorry
-    match ← applyType type.get args with
-    | .sort u => pure u.isZero
+/-- Checks if a type is an unit inductive -/
+def isUnit : Value → TypecheckM Bool
+  | .app _ (.const name i _) _ => do
+    match ← derefConst name i with
+    | .inductive induct => pure induct.unit
     | _ => pure false
-  | .sort _ => pure false
-  | _ => pure false -- these are actually impossible cases
+  | _ => pure false
 
 mutual
   /-- 
-  Checks that `term : Expr` has type `type : Value`, and throws `TypecheckError.valueMismatch`
-  if it is not.
+  Checks if `term : Expr` has type `type : Value`. Returns the expression with flags updated
   -/
-  partial def check (term : Expr) (type : Value) : TypecheckM Unit := do
+  partial def check (term : Expr) (type : Value) : TypecheckM Expr := do
     match term with
     | .lam _ lamName _ _lamDom bod =>
       match type with
-      | .pi _ _ dom img env =>
+      | .pi _ _ _ dom img env =>
+        let lvl := (← read).lvl
+        let var : Thunk Value := {
+          fn := fun _ =>
+            mkVar sorry lamName lvl,
+          repr := "",
+        }
+        let img ← withNewExtendedEnv env var $ eval img
         -- TODO check that `lamDom` == `dom`
         -- though this is wasteful, since this would force
         -- `dom`, which might not need to be evaluated.
-        let var := mkVar lamName (← read).lvl
-        let img ← withNewExtendedEnv env var $ eval img
         withExtendedCtx var dom $ check bod img
```

Comment:
Wouldn't this result in returning the body of the lambda expression rather than the lambda itself? Same with the `.letE` case below.